README.md
# ToyBox-Ext

### Description
Systemless Magisk module
to install (additional) ToyBox applets, using **ARMv7 or higher** binaries from:
https://github.com/landley/toybox

**Supported**: ARMv7 and higher 

##### TEST

Ussually, **ToyBox** comes preinstalled to `/system/bin`, but with limited number of **applets**.
The module installes additional `toybox-ext` binary, with the additional applets (**UNIX/Linux command-line utilities**).

**Install** the module and **reboot**. Use **Terminal Emulator** - **Toybox applets** are for Terminal and **shell scripts**.

Make sure that `/system/bin` and/or `/system/xbin` (the latter might not be available on some devices) are in the `PATH`.
Check e.g. with (all commands are **case sensitive**):

```
su
echo $PATH
```

Check if `toybox-ext` was properly installed and is it installed to `/system/xbin` or `/system/bin`:

```
toybox-ext --help
which toybox-ext
```

If the responded toybox-ext path was `/system/xbin` then check:

```
ls -l /system/xbin | grep toybox-ext
```

whereas if the toybox-ext path was `/system/bin` then use:

```
ls -l /system/bin | grep toybox-ext
```

The command will show all toybox-ext applets as installed (symlinked to the toybox-ext binary).
Usually, `w` will be one of them, hence you can try:

```
w
```

**Note 1**: Some applets are common to **BusyBox** and **ToyBox**, e.g. `ascii`.

Hence, if you have also the `busybox` installed, `toybox-ext` will symlink less number of (additional) applets.

**Note 2**: Since ToyBox-Ext **v1.0.2**, the module also looks if there are not-symlinked applets available from the pre-installed toybox binary (ROMs usually come with the pre-installed ToyBox but with smaller or an older binary and with a less number of applets).

For me, the additionally found applets from the pre-installed ToyBox were e.g. `getfattr` and `iotop`.

### Coffee
If this module is useful for you, please support development by donating a [coffee](https://zgfg.github.io/PayPal.html)

#### Source 

[My Repo:](https://github.com/zgfg/ToyBox-Ext)

![GitHub release (latest by date)](https://img.shields.io/github/v/release/zgfg/ToyBox-Ext?label=Release&style=plastic) ![GitHub Release Date](https://img.shields.io/github/release-date/zgfg/ToyBox-Ext?label=Release%20Date&style=plastic) 
![GitHub Releases](https://img.shields.io/github/downloads/zgfg/ToyBox-Ext/latest/total?label=Downloads%20%28Latest%20Release%29&style=plastic)
![GitHub All Releases](https://img.shields.io/github/downloads/zgfg/ToyBox-Ext/total?label=Total%20Downloads%20%28All%20Releases%29&style=plastic)

[Alt-Repo:](https://github.com/Magisk-Modules-Alt-Repo/ToyBox-Ext)

![GitHub release (latest by date)](https://img.shields.io/github/v/release/Magisk-Modules-Alt-Repo/ToyBox-Ext?label=Release&style=plastic) ![GitHub Release Date](https://img.shields.io/github/release-date/Magisk-Modules-Alt-Repo/ToyBox-Ext?label=Release%20Date&style=plastic) 
![GitHub Releases](https://img.shields.io/github/downloads/Magisk-Modules-Alt-Repo/ToyBox-Ext/latest/total?label=Downloads%20%28Latest%20Release%29&style=plastic)
![GitHub All Releases](https://img.shields.io/github/downloads/Magisk-Modules-Alt-Repo/ToyBox-Ext/total?label=Total%20Downloads%20%28All%20Releases%29&style=plastic)

#### Copyright (c) zgfg @ xda, 2022-



changelog.md
#### Coffee
If the module is useful, you may support development by donating a [coffee](https://zgfg.github.io/PayPal.html)

#### v1.0.8

Service fix for v1.0.7 and improved logging

#### v1.0.7

Logging and notify when new ToyBox binary is ready for installation

#### v1.0.6

Force the latest ToyBox applets 

#### v1.0.5

Update to the latest ToyBox binary on installation 

#### v1.0.4

ToyBox binaries updated to v0.8.9 (build Jan 10, 2023)

#### v1.0.3

ToyBox binaries updated to v0.8.7 (build May 5, 2022)

#### v1.0.2

Install additional stock toybox applets if found

#### v1.0.1

Alt Repository release

#### v1.0.0

Find the most suitable ARMv7 or higher toybox binary

#### v0.9.9

Initial release - Install (additional) toybox applets. SUPPORTED: ARMv7 and higher 


service-test.sh
#!/system/bin/sh

# Magisk Module: ToyBox-Ext v1.0.8
# Copyright (c) zgfg @ xda, 2022-
# GitHub source: https://github.com/zgfg/ToyBox-Ext

# Module's own path (local path)
MODDIR=${0%/*}
cd $MODDIR

# Log for debugging
LogFile="$MODDIR/service.log"
exec 3>&2 2>$LogFile
set -x
LogTime=$(date +%c)

# Current time
DLTIME=$(date +"%s")

# Source the original toybox binary type and last download time
TBSCRIPT='./tbtype.sh'
if [ -f $TBSCRIPT ]
then
  . $TBSCRIPT
fi

# Passed time since the last download
PASSEDTIME=$(($DLTIME - $LASTDLTIME))

# Waiting time between downloads (15 days)
#WAITTIME=$((15 * 24 * 3600))
WAITTIME=$((20  * 60))  # toDo: Remove - for testing

# If waiting time passed, download the latest binary again
if [ ! -z $TBTYPE ] && [ $PASSEDTIME -gt $WAITTIME ]
then
  # Wait to finish booting
  until [ "$(getprop sys.boot_completed)" = 1 ]
  do
    sleep 1
  done

  # and few more seconds
  sleep 3
  rm -f $TBTYPE
  /data/adb/magisk/busybox wget -c -T 20 "http://landley.net/toybox/bin/$TBTYPE"
fi

# Test the download 
if [ ! -z $TBTYPE ] && [ -f $TBTYPE ]
then
  # Compare checksums for the old and new binary
  MD5Old=$(md5sum toybox-ext | head -c 32)
  MD5New=$(md5sum "$TBTYPE" | head -c 32)
  if [ "$MD5New" = "$MD5Old" ]
  then
    # Save the download time
    echo "LASTDLTIME=$DLTIME" >> $TBSCRIPT

    # Delete, same as old binary
    rm -f $TBTYPE
  else
    # Test downloaded binary
    chmod 755 $TBTYPE
    Applets=$(./$TBTYPE)
    if [ -z "$Applets" ]
    then
      # Delete, not working
      rm -f $TBTYPE
    else
      # Save the binary type and installation time
      echo "TBTYPE=$TBTYPE" > $TBSCRIPT
      echo "LASTDLTIME=$DLTIME" >> $TBSCRIPT

      # Notify user to reboot
      exec 2>&3 3>&-
      su -lp 2000 -c "cmd notification post -S bigtext -t 'ToyBox-Ext Module' 'Tag' 'Reboot to update ToyBox binary'" 1>/dev/null
	  exec 3>&2 2>>$LogFile
    fi
  fi
fi

customize.sh
#!/system/bin/sh

# Magisk Module: ToyBox-Ext v1.0.8
# Copyright (c) zgfg @ xda, 2022-
# GitHub source: https://github.com/zgfg/ToyBox-Ext

if [ -z $BOOTMODE ] ||  [ "$BOOTMODE" != "true" ] 
then
	abort "ERROR: Install from Magisk app, not from TWRP!"
fi

# Module's own path (local path)
cd $MODPATH

# toybox ARMv7 and higher binaries
TBTYPEList="
toybox-aarch64
toybox-armv7m
toybox-armv7l
"

echo "ABI: $(getprop ro.product.cpu.abi)"
echo "ABILIST: $(getprop ro.product.cpu.abilist)"

# toybox binary to be installed
TBEXT=toybox-ext

# Find the applicable binary
TBTYPE=""
for TB in $TBTYPEList
do
  if [ -z $TBTYPE ]
  then
    # Test if binary executes 
    echo "Testing archived $TB"
    chmod 755 $TB
    Applets=$(./$TB)

    if [ ! -z "$Applets" ]
    then
      # Applicable binary found
      TBTYPE=$TB
      mv $TBTYPE $TBEXT
      echo "Archived $TBTYPE installed"
      continue
    fi
  fi

  # Delete binary (already found or doesn't execute)
  rm -f $TB
done

if [ -z $TBTYPE ]
then
  # Applicable binary not found
  echo
  echo "ERROR: ToyBox not installed!"
  echo
  echo "$(cat /proc/cpuinfo)"
  echo
  exit -1
fi

# Download latest binary
echo "Downloading latest $TBTYPE"
wget -c -T 10 "http://landley.net/toybox/bin/$TBTYPE"

# Test the download 
if [ ! -f $TBTYPE ]
then
  # Not downloaded 
  echo "$TBTYPE not downloaded"
else
  echo "Testing downloaded $TBTYPE"
  chmod 755 $TBTYPE
  Applets=$(./$TBTYPE)
  if [ ! -z "$Applets" ]
  then
    # Install 
    mv $TBTYPE $TBEXT
    echo "Downloaded $TBTYPE installed instead"
  else
    # Delete, not working
    echo "Use archived $TBTYPE instead"
    rm -f $TBTYPE
  fi
fi

# Current time
DLTIME=$(date +"%s")

# Save the toybox binary type and installation time
TBSCRIPT='./tbtype.sh'
echo "TBTYPE=$TBTYPE" > $TBSCRIPT
echo "LASTDLTIME=$DLTIME" >> $TBSCRIPT


service-log.sh
#!/system/bin/sh

# Magisk Module: ToyBox-Ext v1.0.7
# Copyright (c) zgfg @ xda, 2022-
# GitHub source: https://github.com/zgfg/ToyBox-Ext

# Wait to finish booting
until [ "$(getprop sys.boot_completed)" = 1 ]
do
  sleep 1
done

# Module's own path (local path)
MODDIR=${0%/*}
cd $MODDIR

# Current time
DLTIME=$(date +"%s")

# Source the original toybox binary type and last download time
TBSCRIPT='./tbtype.sh'
if [ -f $TBSCRIPT ]
then
  . $TBSCRIPT
fi

# Passed time since the last download
PASSEDTIME=$(($DLTIME - $LASTDLTIME))

# Waiting time between downloads (15 days)
WAITTIME=$((15 * 24 * 3600))

# If waiting time passed, download the latest binary again
if [ ! -z $TBTYPE ] && [ $PASSEDTIME -gt $WAITTIME ]
then
  sleep 5
  rm -f $TBTYPE
  /data/adb/magisk/busybox wget -c -T 10 "http://landley.net/toybox/bin/$TBTYPE"
fi

# Test the download 
if [ ! -z $TBTYPE ] && [ -f $TBTYPE ]
then
  # Compare checksums for the old and new binary
  MD5Old=($(md5sum toybox-ext))
  MD5New=($(md5sum "$TBTYPE"))
  if [ "$MD5New" = "$MD5Old" ]
  then
    # Delete, same as old binary
    rm -f $TBTYPE
  else
    # Test downloaded binary
    chmod 755 $TBTYPE
    Applets=$(./$TBTYPE)
    if [ -z "$Applets" ]
    then
      # Delete, not working
      rm -f $TBTYPE
    else
      # Save the binary type and installation time
      echo "TBTYPE=$TBTYPE" > $TBSCRIPT
      echo "LASTDLTIME=$DLTIME" >> $TBSCRIPT

      # Notify user to reboot
      su -lp 2000 -c "cmd notification post -S bigtext -t 'ToyBox-Ext Module' 'Tag' 'Reboot to update ToyBox binary'" 1> /dev/null
    fi
  fi
fi


post-fs-data.sh
#!/system/bin/sh

# Magisk Module: ToyBox-Ext v1.0.8a
# Copyright (c) zgfg @ xda, 2022-
# GitHub source: https://github.com/zgfg/ToyBox-Ext

# Module's own path (local path)
MODDIR=${0%/*}

# Log file for debugging
LogFile="$MODDIR/post-fs-data.log"
exec 2>$LogFile 1>&2
set -x

# Log Magisk version and magisk --path
magisk -c
magisk --path

# Source the original toybox binary type
cd $MODDIR
pwd
TBSCRIPT='./tbtype.sh'
if [ -f $TBSCRIPT ]
then
  . $TBSCRIPT
fi

# ToyBox-Ext
TB=toybox
TBEXT=toybox-ext
if [ ! -z $TBTYPE ] && [ -f $TBTYPE ]
then
  # Replace toybox-ext with the latest downloaded binary
  mv $TBTYPE $TBEXT
fi
chmod 755 $TBEXT

# Clean-up old stuff
rm -rf "$MODDIR/system"

# Choose XBIN or BIN path
SDIR=/system/xbin
if [ ! -d $SDIR ]
then
  SDIR=/system/bin
fi
TBDIR=$MODDIR$SDIR
mkdir -p $TBDIR
cd $TBDIR
pwd

# ToyBox-Ext applets
TBBIN=$MODDIR/$TBEXT
Applets=$TB$'\n'$TBEXT$'\n'$($TBBIN)

# Create symlinks for toybox-ext applets
for Applet in $Applets
do
  # Skip if applet already found in the path
  Check=$(which $Applet)
  if [ -z "$Check" ] && [ ! -x "$SDIR/$Applet" ]
  then
    # Create symlink
    ln -s $TBBIN $Applet
  fi
done

# Stock ToyBox applets
TBSTOCK=$(which $TB)
Applets=""
if [ ! -z "$TBSTOCK" ]
then
  TBPATH=$(echo "$TBSTOCK" | sed "s,/$TB$,,")
  Applets=$TB$'\n'$($TBSTOCK)
fi

# Create symlinks for toybox-stock applets
for Applet in $Applets
do
  # Skip if applet already found in the path
  Check=$(which $Applet)
  if [ -z "$Check" ] && [ ! -x "$SDIR/$Applet" ] && [ ! -x "$TBPATH/$Applet" ] && [ ! -x "$Applet" ]
  then
    # Create symlink
    ln -s $TBSTOCK $Applet
  fi
done

# Log results for ToyBox-Ext
ls -l $TBEXT
$TBBIN --version
ls -l | grep $TBEXT | grep ^lr.x | rev | cut -d ' ' -f 3 | rev
ls -l | grep $TBEXT | grep ^lr.x | wc -l

# Log results for stock ToyBox
$TBSTOCK --version
ls -l | grep $TB | grep -v $TBEXT | grep ^lr.x | rev | cut -d ' ' -f 3 | rev
ls -l | grep $TB | grep -v $TBEXT | grep ^lr.x | wc -l


service.sh
#!/system/bin/sh

# Magisk Module: ToyBox-Ext v1.0.8a
# Copyright (c) zgfg @ xda, 2022-
# GitHub source: https://github.com/zgfg/ToyBox-Ext

# Module's own path (local path)
MODDIR=${0%/*}

# Log for debugging
LogFile="$MODDIR/service.log"
exec 3>&2 2>$LogFile 1>&2
set -x
date +%c

# Log Magisk version and magisk --path
magisk -c
magisk --path

# Log results for ToyBox-Ext
TBEXT=toybox-ext
which $TBEXT
$TBEXT --version
SDIR=$(which $TBEXT | sed "s,/$TBEXT$,,")
if [ -d $SDIR ]
then
  cd $SDIR
  pwd
  ls -la | grep $TBEXT | grep ^lr.x | rev | cut -d ' ' -f 3 | rev
  ls -la | grep $TBEXT | grep ^lr.x | wc -l
fi

# Log results for stock ToyBox
TB=toybox
which $TB
$TB --version
SDIR=$(which $TB | sed "s,/$TB$,,")
if [ -d $SDIR ]
then
  cd $SDIR
  pwd
  ls -la | grep $TB | grep -v $TBEXT | grep ^lr.x | rev | cut -d ' ' -f 3 | rev
  ls -la | grep $TB | grep -v $TBEXT | grep ^lr.x | wc -l
fi

# Source the original toybox binary type and last download time
cd $MODDIR
pwd
TBSCRIPT='./tbtype.sh'
if [ -f $TBSCRIPT ]
then
  . $TBSCRIPT
fi

# Current time
DLTIME=$(date +"%s")

# Passed time since the last download
PASSEDTIME=$(($DLTIME - $LASTDLTIME))

# Waiting time between downloads (15 days)
WAITTIME=$((15 * 24 * 3600))

# If waiting time passed, download the latest binary again
if [ ! -z $TBTYPE ] && [ $PASSEDTIME -gt $WAITTIME ]
then
  # Wait to finish booting
  until [ "$(getprop sys.boot_completed)" = 1 ]
  do
    sleep 1
  done

  # and few more seconds
  sleep 3
  rm -f $TBTYPE
  /data/adb/magisk/busybox wget -c -T 20 "http://landley.net/toybox/bin/$TBTYPE"
fi

# Test the download 
if [ ! -z $TBTYPE ] && [ -f $TBTYPE ]
then
  # Compare checksums for the old and new binary
  MD5Old=$(md5sum toybox-ext | head -c 32)
  MD5New=$(md5sum "$TBTYPE" | head -c 32)
  if [ "$MD5New" = "$MD5Old" ]
  then
    # Save the download time
    echo "LASTDLTIME=$DLTIME" >> $TBSCRIPT

    # Delete, same as old binary
    rm -f $TBTYPE
  else
    # Test downloaded binary
    chmod 755 $TBTYPE
    Applets=$(./$TBTYPE)
    if [ -z "$Applets" ]
    then
      # Delete, not working
      rm -f $TBTYPE
    else
      # Save the binary type and installation time
      echo "TBTYPE=$TBTYPE" > $TBSCRIPT
      echo "LASTDLTIME=$DLTIME" >> $TBSCRIPT

      # Notify user to reboot
      exec 2>&3 3>&-
      su -lp 2000 -c "cmd notification post -S bigtext -t 'ToyBox-Ext Module' 'Tag' 'Reboot to update ToyBox binary'" 1>/dev/null
	  exec 3>&2 2>>$LogFile 1>&2
    fi
  fi
fi
